Support tensor subclass unwrapping

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #143671
* __->__ #141941


This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D67531057](https://our.internmc.facebook.com/intern/diff/D67531057)

tugsbayasgalan committed:
Support tensor subclass unwrapping

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66690419).

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/custom_tensor.py:
```diff
@@ -65,3 +65,81 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):
+        # Don't use this tensor with view ops
```
why not? (if we don't want to add view support for now, you can raise an explicit `NotImplementedError` if `func.is_view`)

tugsbayasgalan replied:
Resolved offline.

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented on test/export/test_export.py:
```diff
@@ -1415,6 +1415,63 @@ def foo(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        ref_x = torch.randn(3, 4)
+        ref_out = m(ref_x)
+
+        ep_training = torch.export.export_for_training(m, (ref_x,))
```
One thing that would be nice for E2E testing is to confirm that this actually works with AOTI. wdyt?

tugsbayasgalan replied:
Yep added!

tugsbayasgalan commented on torch/testing/_internal/custom_tensor.py:
```diff
@@ -65,3 +65,82 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
+
+# A simple tensor subclass that always returns plain tensor during __torch_dispatch__
+# It is similar to TwoTensor and is used to simulate torchao quantized tensors
+class CustomTensorPlainOut(torch.Tensor):
```
Yep good idea

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66690419).

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/compile_fx.py:
```diff
@@ -1543,6 +1547,9 @@ def compile_fx(
     """
+    
+    # [See NOTE] Unwrapping subclasses AOT
+    unwrap_tensor_subclass_parameters(model_)
```
huh... this will run whenever `compile_fx()` is called, which is basically the top-level function invoked in inductor when we run `torch.compile`. Why do we want it here?

bdhirsh replied:
I was imaging more that once you have `unwrap_tensor_subclass_parameters` getting implicitly used by `export.run_decompositions()`, then in theory full AOTI support should come for free - since the function that we are passing to inductor in the AOTI case has already been parametrized. So we just need some testing that the E2E case works properly.

Since torchAO are some of the main customers, we probably want to make sure their export cases work automatically with AOTI too

tugsbayasgalan replied:
Yup! I talked with @desertfire  bit about this, so the aot_export_module still exist in aot_inductor because originally aoti code was designed to handle any fx graph (including export) hence this fork exist. But we both agreed that in the long term, run-decomp should consume this part.

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/compile_fx.py:
```diff
@@ -1326,6 +1329,9 @@ def compile_fx_aot(
     assert isinstance(model_, GraphModule), model_
 
+    # [See NOTE] Unwrapping subclasses AOT
+    unwrap_tensor_subclass_parameters(model_)
```
cc @angelayi @desertfire. I think this is a case where `aot_compile()` and `ExportedProgram.run_decompositions()` are doing some similar things but they aren't sharing much code, so right now we're duplicating some code between the two (in this case, handling tensor subclass weights during export). Maybe in the future we can think about a way to refactor/share more code?

bdhirsh reviewed (approved):
also looks like there are still some test failures (giving a preemptive stamp)

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclass unwrapping"


Differential Revision: [D66690419](https://our.internmc.facebook.com/intern/diff/D66690419)

This PR adds support for export to unwrap/wrap subclasses AOT so that we can trace through subclass parameters. This will resolve the UX issue in torchao where users had to manually unwrap their subclasses before calling export.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67531057).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67531057).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67531057).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67531057).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67531057).
